### ISSUE TITLE
Lua: Strategy for LSP configuration

### ISSUE THREAD

**Issue by tjdevries** [neovim/neovim#12277]
After reviewing some PRs and hearing people discuss various aspects of using LSP, I've been trying to figure out how we can make the experience more easily customizable (and particularly, without having to copy and paste the source and replace one line).

Some strategies:
1. As linked here: https://github.com/neovim/neovim/pull/12257#issuecomment-625260364 , expose more functions, even as "private" functions that people can override to change behavior slightly.
2. Develop some sort of LSP config (whether through global variables [eww] or some other method). I think this would be useful anyway in a more structured way because a lot of people have similar complaints about a few things (like updating diagnostics while in insert mode always).
3. Others?

@teto @justinmk wondering what your thoughts are for this or if I missed something we've discussed elsewhere.

cc @clason

**Comment by clason** [comment]
> 1. As linked here: [#12257 (comment)](https://github.com/neovim/neovim/pull/12257#issuecomment-625260364) , expose more functions, even as "private" functions that people can override to change behavior slightly.

The main argument in that comment was actually the other way around: if I want to override a utility function, I may not want to override all the local functions that are used there as well. At the moment I have to look for all the local functions used and copy them into my own lua file; if they were exposed, I could just replace every `M.xxx` with, say, `util.xxx` in the overriden function and call it a day. (And, more importantly, profit from bugfixes in core.)

**Comment by teto** [comment]
I am very happy with diagnostic-nvim so it is possible to get pretty far with the current interfaces.
As a first step it's good enough to provide a common server description interface that can be leveraged by coc, lsp-vim, LanguageClient etc.
The out of the box experience could be improved for sure but I would postpone any commitment to a more advanced interface to another release such as 0.5.1.

**Comment by clason** [comment]
> I am very happy with diagnostic-nvim so it is possible to get pretty far with the current interfaces.

The point is relevant here, too: If you look at the actual code, you see a _lot_ of functions there that are just copied (without changes) from the built-in code (where they are local). This brings with it the danger of bitrot (every bugfix and improvement has to be backported manually -- not just to this but to all other plugins that build on core). Vice versa, bugfixes to shared code end up in the plugin and not in core.

I'd argue that allowing them to (optionally)  reuse _any_ internal function would lead to a better experience. (Not going to push it further, though.)

**Comment by mfussenegger** [comment]
> This brings with it the danger of bitrot (every bugfix and improvement has to be backported manually -- not just to this but to all other plugins that build on core). Vice versa, bugfixes to shared code end up in the plugin and not in core.

The downside of this is that any change in neovim that changes the signatures of those functions may end up breaking plugins.

I think for the long-term health of the code base - from a maintenance perspective -  it is beneficial to keep being restrictive with access to internals.

Let plugin authors continue to explore what kind of functionality is useful. If that involves copy and paste so be it. The low-level building blocks with `buf_request` are already incredibly powerful.

Then decide on a use-case basis whether something should be in the core to begin with, or whether it makes sense to expose functions that were not exposed in the beginning. Keep YAGNI in mind.

Making changes to LSP parts already starts becoming more painful because of BWC.

**Comment by tjdevries** [comment]
I guess there are two aspects I'm looking to discuss, probably should have made it more clear.

1. Provide some sort of configuration options (via global, global dictoinary, functions, json, ???, etc.) so that people can do a few things more easily (choose which autocmds fire for diagnostics, for example).

2. Make the code more composable so you can override certain aspects of the builtin LSP.

For example, we recently merged something and used `vim.fn.inputlist` which is a fine choice. But we could have chosen to do something like make a function `util.select_input_from_strings` or make it private via `util._select_input_from_strings` and allowed people to easily override that by patching over that function (maybe they want to use FZF or denite).

I think we could expose more in a way that says "this is private and we make no guarantees it will stay working", which the convention has been preceding underscore. This seems in some way to be a good compromise (at least initially) because we get to change those things whenever we want but if the adventurous plugin developer / personal user wants to override some aspect, they can. Over time, as we're confident that people are using X or Y, we can move them from private to public and then be concerned w/ BWC.

I don't think we should expose everything, but perhaps we can keep an eye out for composability while developing and reviewing PRs so that it can allow people to customize their experience even more so. That is one thing I love about nvim. I can spend more time customizing than working ;)

But you do bring up some very good points @mfussenegger . I think we have already exposed very powerful building blocks and people are using them in exciting ways :)

**Comment by haorenW1025** [comment]
Just some of my opinion about easier lsp config. Is that possible to create a `lsp_config.lua` and put the configuration option in there as a table? Something like
```lua
M.lsp_config_option = {
  diagnostic_virtual_text = true
  diagnostic_sign = true
...
}
```
Users can overwrite this by copy this file into `~/.config/nvim/lua` and overwrite it. This should at least be much better than having to copying and delete lines of code from source and modified them.

**Comment by tjdevries** [comment]
@haorenW1025 Yes. I think we should do something similar to that. I have proposed in https://github.com/neovim/neovim/pull/12333 a way to make the configuration better than simply a table (as it's easy to mistype the table, have your table go out of date, change the value and not have it update, etc.).

This should allow us to do things like set autocmds via the "callback" or "side effect" functions in the configuration, so we don't have to litter those kind of configuration options all over the place.

**Comment by landerlo** [comment]
I'm contributing to @ckipp01's nvim-lsp port of his coc-metals for the popular scala 'metals' LSP server. First of all thank you for the fantastic work done with nvim-lsp. I have done a small but fully functional POC to showcase an approach that I think will help with the configuration scenarios user will come up with: https://github.com/neovim/neovim/pull/12347

I'm first focusing on the location queries, ```textDocument/definition...references```, etc. Many user configuration needs will be regarding these queries as they are an important part of the UI.

I like the power of quicklists but I also think other options with previews and fuzzy search eye candy will be important to attract users from coc.nvim and vscode. I personally use vim#fzf for fuzzy selecting like preview, as it also allows to select entries to to go to cf if I want to cdo. IMO the configurability here is key. We already see some challenges like the peek_defintion request: https://github.com/neovim/neovim/pull/12232

My proposal is to accept high level callbacks as arguments to this functions if we want to override the default UI. For the location queries, the callback is function(locations) that gets the list of locations in case of no errors or locations available. We removed the error cases and sanitized the differences were location can be a string or a list. The POC is fully functional and allows my config to be:

```vim
nnoremap <silent> gd          <cmd>lua vim.lsp.buf.definition()<CR>
if exists('fzf')
  nnoremap <silent> gD        <cmd>lua vim.lsp.buf.definition { on_locations=metals.fuzzy_select }<CR>
endif
nnoremap <silent> gr          <cmd>lua vim.lsp.buf.references { includeDeclaration = true }<CR>
nnoremap <silent> gp          <cmd>
  \lua vim.lsp.buf.definition { on_locations= function(locs) vim.lsp.util.preview_location(locs[1]) end } <CR>
```
It fullfils the peek request https://github.com/neovim/neovim/pull/12232 by simply copying the preview_location function, without any of the other extra complexity of extensions and the potential to having to rewrite functions.

This makes the configuration for these RPC requests very vim like and declarative when using lua tables like in my working config example. As a bonus they don't need interaction to the setup function and the code is simpler and more localised.

I think it's also clearer to separate, both in code and in terminology, what is esentially handlers for server initiated request, e.g. like workspaceEdits or diagnostics, from the async callbacks to process the server responses to the client requests. The request handlers need to be attached and registered. On the other hand the callbacks thar process the reponse when the client sends a request, those can be passed at call time allowing the extra flexibility and clarity with IMHO no downsides, other than some extra complexity of the function composition technique. But I think it's worth it to reap the benefits of the great foundations of buf_request, and the underlying neovim infrastructure and lua engine. It really makes it shine.

**Comment by mfussenegger** [comment]
> My proposal is to accept high level callbacks as arguments to this functions if we want to override the default UI. For the location queries, the callback is function(locations) that gets the list of locations in case of no errors or locations available.

> On the other hand the callbacks thar process the reponse when the client sends a request, those can be passed at call time allowing the extra flexibility and clarity with IMHO no downsides, other than some extra complexity of the function composition technique.


To me it seems that this solution would be some kind of middle-ground to the current state. It's neither a low level API, nor high level. Having disadvantages from both. It narrows what kind of customization can be done, although that specific customization will be easier. 

I also fear that such an approach would be prone to leaking implementation details into the API - making it harder to change things in the future.

---

To recap:

1. The current request/callback model is easy to understand, and allows both users and plugins to customize pretty much anything. I think this model is working really well and don't think the overall approach should be changed.


2. Customization can involve quite a bit of boilerplate


I think there are several aspects and questions here:


- What customizations and options do users even want? One common thing that came up so far is changing the behavior of the diagnostics (which is why https://github.com/haorenW1025/diagnostic-nvim got created)

- Which customizations should be done by plugins, which should have options? My opinion here is that in a first step this could be delegated mostly to plugins - it has already happened anyway, and then we could look at what options users commonly set and introduce this into the neovim core. 
For those that should have options, I think the config table/option approach referenced looks good.

- What can be done to make these customizations easier? 

One thing that comes to mind here is that maybe this question shouldn't focus on LSP alone, but also think outside of the LSP box and consider neovim extension in general. 
Just as an example as fzf over quickfix-list was mentioned: What if neovim provided an extension point independent of LSP, that allows to tweak or even override the quickfix list?

Another thought: There is quite some complexity in the LSP utils section because some of the nvim APIs don't fit well with the structures used in the language server protocol. One example is the range/position conversion (handling byteindex). Maybe what's missing and what leads to a lot of boilerplate / code copying is a lack of APIs that fit better with the given structures. Some of that is already being resolved (e.g. what is happening with a `nvim_buf_set_text` method).
Maybe we can put some effort into extracting some more low-level functions - which would be useful outside of the LSP scope) - before exposing all the high and mid-level util methods.

**Comment by landerlo** [comment]
> 
> To me it seems that this solution would be some kind of middle-ground to the current state. It's neither a low level API, nor high level. Having disadvantages from both. It narrows what kind of customization can be done, although that specific customization will be easier.
> 
> I also fear that such an approach would be prone to leaking implementation details into the API - making it harder to change things in the future.

I'd be great if you could elaborate on the downsides and the leaking API fears. Just to clarify, I'm not proposing to change what's already there for diagnostics, edits, etc. Those work great and they need to be attached and registered as they are now.

I'm only focusing on the interactions started by the client that follow the RPC client request -> server response (locations, available actions, symbols) -> choice of action by the user. These involve UI and are very likely to be customised, and more importantly, can have more than one response. Peek is a great example.  One binding could call textDocument/definition and peek on the returned location and another does the jump/qf, etc. And the request is the same textDocument/definition, we cannot have a single callback for this request, and having to go lower level to have more than one callback it's going to create issues and difficulties for the user.

There is a lot of value to have these 'configured' on the binding call site, so they are orthogonal to the plugin setup and don't have interactions with them. The user can choose what they want and we just provide the util functions that compose with the reponse, i.e. jump_to_location or preview_location.
Simpler configuration and less error prone than the user having to register callbacks in the setup function with more possibilities of error and misconfiguration.

For things like workspaceEdit what we have works great, the user expects the edits to just work and I cannot think of a way to apply them differently. It'll be only advanced users or plugins wanting to chane it and it will require careful thought anyway so the ease of use considerations are less relevant and there is safety in keeping the configurability more hidden from the average user and low level.

As an example for the metals LSP server my ideal would be to have Plug nvim/metals, diagnostics and completions, work out of the box without having to configure anything if the lsp server executable is in the path and I don;t want to change the defaults. It does the setup or provides a setup_nvim_metals( {overrides}). Then i can configure UI element for the choices and for the actions on locations codeActions in my normal key bindings.  

 
> 
> To recap:
> 
>     1. The current request/callback model is easy to understand, and allows both users and plugins to customize pretty much anything. I think this model is working really well and don't think the overall approach should be changed.
> 
>     2. Customization can involve quite a bit of boilerplate
> 
> 
> I think there are several aspects and questions here:
> 
>     * What customizations and options do users even want? One common thing that came up so far is changing the behavior of the diagnostics (which is why https://github.com/haorenW1025/diagnostic-nvim got created)
> 
>     * Which customizations should be done by plugins, which should have options? My opinion here is that in a first step this could be delegated mostly to plugins - it has already happened anyway, and then we could look at what options users commonly set and introduce this into the neovim core.
>       For those that should have options, I think the config table/option approach referenced looks good.
> 
>     * What can be done to make these customizations easier?
> 
> 
> One thing that comes to mind here is that maybe this question shouldn't focus on LSP alone, but also think outside of the LSP box and consider neovim extension in general.
> Just as an example as fzf over quickfix-list was mentioned: What if neovim provided an extension point independent of LSP, that allows to tweak or even override the quickfix list?
> 
> Another thought: There is quite some complexity in the LSP utils section because some of the nvim APIs don't fit well with the structures used in the language server protocol. One example is the range/position conversion (handling byteindex). Maybe what's missing and what leads to a lot of boilerplate / code copying is a lack of APIs that fit better with the given structures. Some of that is already being resolved (e.g. what is happening with a `nvim_buf_set_text` method).
> Maybe we can put some effort into extracting some more low-level functions - which would be useful outside of the LSP scope) - before exposing all the high and mid-level util methods.

**Comment by clason** [comment]
Having been trained by Stackexchange never to downvote without leaving a comment, here's my take on this point (which is kind of orthogonal to the initial question of exposing more low-level API, where I agree with your comment that this should be done in a more considered fashion than merely making _everything_ public).

First, like @landerlo, I feel that your recap is a bit unfair and one-sided and requires a bit of elaboration. Second, I would argue that a clear separation of client--server communication and response handling is in the spirit of the original proposal (making a powerful lsp library with reasonable out-of-the-box defaults). Polemically turning tables for a moment, the approach you seem to advocate would lead to users having to install `definition.nvim`, `typeDefinition.nvim`, `references.nvim`... just to (say) open the target in a new split. Another use case of @landerlo's approach is to turn on wrapping in the `hover` window -- which I think metals does, and so do I in my config.

(Completion.nvim is kind of different since what it does is in principle independent of LSP.)

**Comment by mfussenegger** [comment]
> I'd be great if you could elaborate on the downsides and the leaking API fears. Just to clarify, I'm not proposing to change what's already there for diagnostics, edits, etc. Those work great and they need to be attached and registered as they are now.

Let me try:

I like to split use-cases into the 90-95% majority case, and then the remaining 5% special cases.

In the neovim case, I'd think that the ~90% users probably don't want to customize too much. They've a `init.vim` where they set a couple of option that they found useful. They treat it as a config file and don't want to write code.

For those, configuration options like `set lsp_diagnostics-=virtualtext` or something like that - options that feel like other "vim" options - would probably work best. (Similar to how `completeopt` and other stuff works). Using Lua config tables may work for them too.

The goal for them should be to make the out-of-the-box experience good, and to provide options where the preferences have a wide range or where the behavior is controversial. (E.g. the diagnostics being too aggressive popped up a lot).
If people now want to customize X, the question shouldn't just be whether how to make X easier to customize, but also ask why they want to do that, and consider improving the out-of-box behavior. 
For example I think the `hover` output has been customized by a few, probably because the formatting of the current built-in could use some tweaks.

These users are likely also comfortable with adding plugins, for the cases where the out-of-the-box experience is not yet good enough.



Then you've the remaining ~5-10%, with a very wide range of requirements and preferences, including plugin authors. Trying to provide extension points and configuration option for each and every use-case is a difficult task. That's why the low-level API is so powerful, it is generic and allows pretty much any customization.


The proposed POC (https://github.com/neovim/neovim/pull/12347) has a couple of use cases in mind. One to open the location elsewhere. So an option is added to customize that. The next user will ask to be able to add the last error message to the status line, so there will be an option for that. Then somebody wants to keep it in the quickfix list, but without opening the quickfix list. So there will be an option for that. It's `n+1` to accommodate all imaginable use-case. Each of which can already be accomplished by using `vim.lsp.buf_request(0, method, params, callback)`.

(I'm aware I'm exaggerating here, but I hope the point I'm trying to make is understood)

Using `vim.lsp.buf_request(0, method, params, callback)` directly requires more boilerplate, but I think in the end it is *simpler* (not *easier*) to understand. More importantly, it keeps the API narrow and reduces complexity.

Introducing configuration options to functions that don't influence the behavior, but rather change the behavior also tends to lead to functions like this:

```
if user_wants_behavior_a then
  do_a()
elseif user_wants_behavior_b then
  do_b()
...
```

Which to me is a sign that there really should be multiple functions instead.


In any case, this is just my perspective/feedback. With software there is often no right or wrong, just different trade-offs and different people value different things.

**Comment by ckipp01** [comment]
If I'm able to intermingle a couple question with some comments.

> As a first step it's good enough to provide a common server description interface that can be leveraged by coc, lsp-vim, LanguageClient etc.

Is this actually a goal? I also wasn't around when the initial LSP conversations were happening, but I never heard this before, and didn't assume that this was actual the case. Like do we know that coc.nvim is intending to use the interface for LSP that nvim provides? My naive thought on that is that there is no way they are going to. Any insight on this?

My biggest point probably goes to the initial question posed here by @tjdevries 

> Develop some sort of LSP config (whether through global variables [eww] or some other method). I think this would be useful anyway in a more structured way because a lot of people have similar complaints about a few things (like updating diagnostics while in insert mode always).

At this point this is my biggest concern, and to broaden the conversation a bit, I also think that nvim-lsp should be brought into this conversation since that's where for most basic users this all starts. I know for [Metals](https://scalameta.org/metals/) for example we need to have a base configuration that is currently living in nvim-lsp that works for a user that wants the absolute bare minimum, but then if we want to actually support the other huge chunk that Metals has to offer (different detection of the root workspace, LSP extensions, commands to import the build, etc), then they need to use [nvim-metals](https://github.com/scalameta/nvim-metals). When they do that, a new set of problems arise. The user has to override some of the `init_options`, override some callbacks, and add some new callbacks in to support the LSP extensions. Some of that can be simplified and is possibly my crappy design, but if we then also introduce a config table like is outlined here https://github.com/neovim/neovim/pull/12333, then we now have like 3 different ways the user has to set stuff in order to get their LSP experience to work as desired. I know some of this will probably be smoothed out, but I think that is going to kill the user experience. Many users from coc for example are going to expect to install something, and then have everything just work.

To summarize this up in a couple points:

1. I think having a centralized place for _user_ configuration is key, whether it's all contained in nvim-lsp, only in a configuration table, etc.
2. I agree with @landerlo that the more UI centric settings are important. At times if someone is overriding a callback due to something small, they also may be conflicting with other plugins. As an example I override hover to get the text to wrap, but that conflicted with completion-nvim https://github.com/haorenW1025/completion-nvim/issues/63. So the more plugins that get created, I'm curious how they will all start to work together if multiple are attempting to override the same thing.

**Comment by clason** [comment]
My impression from the development around last year's end (and the regulars may correct me on that) as well as the experience with the lsp functionality:

1. The goal was to get a flexible, performant base infrastructure for lsp in together with a bare-bones but functional frontend that is more a proof of concept than a turn-key solution. Then came norcalli/nvim-common-lsp, and then some posts on reddit, and then people got excited...
2. coc-nvim is based on VS Code extension and therefore a completely different ecosystem. On the other hand, vim-lsp and Language-Client make significant efforts to support both neovim and vim (in all the versions). So neither is very likely to change, especially if they aim to support older versions of neovim too.
3. So the main audience (at least in my view at that time) is people implementing file-type or, rather, server-specific plugins (exactly like @ckipp01 is doing, or a texlab-nvim).
4. However, after spending time with it, I find the design of the lsp functionality is inspired and so easy to get into that I would say @mfussenegger's 90/10 split is missing a sizeable percentage of people who'd never think of putting more than a dozen `let g:whatever = 1` in their `.vimrc` before that are now considering hacking lua to get the functionality _just right_. (Case in point: me, who knew zero lua and wouldn't touch vimscript with a ten-pole stick if possible.) Separating backend from frontend would lower the bar to get more people there (maybe from 10 to 20%? I don't know). And everyone of them potentially becomes a neovim contributor...
5. Regarding configuration, I think the separation is clear: _server_ options go in the setup function provided by `nvim-lsp` (soon to be renamed, hopefully, and stripped of utility functions that are moved to core); `client` options go into the proposed table from #12333.

**Comment by landerlo** [comment]
@mfussenegger I agree with the approach of limiting variability, but the POC doesn't compromise on any of those goals. On the contrary, by being more composable and not interacting with the setup and the registering of other callbacks it's IMO a simpler solution that reduces some complexity.

> The `peekDefinition` case was mentioned. Now you may add that as a callback to a function, in which case you tell the user about the implementation: This function fetches definitions, you can then customize how and where to display it. vs. having two different functions, because for a end-user peeking a definition and listing the definitions are not necessarily related.

This is the problem, you suggest to have two functions. For this case we cannot have two functions because the registered callback is unique, so you propose that to have another function we have to go to a lower level and copy paste code, I don't see the need for that in cases like this as showcased in the PR.

As it's becoming the theme, we need some kind of shared configuration, but for a subset of cases like the location query cases covered in my PR the high level callback passed as an argument works well. More importantly, it doesn't interact with the setup function and the callback registering. I'm hammering again on the same point, but we really need to remove as many configuration moving parts as possible whenever we have a composable and orthogonal solution that doesn't incur in compromises or more complexity.

**Comment by ckipp01** [comment]
> Regarding configuration, I think the separation is clear: server options go in the setup function provided by nvim-lsp (soon to be renamed, hopefully, and stripped of utility functions that are moved to core); client options go into the proposed table from #12333.

Sure, I didn't mean to imply the separation wasn't clear, but a lot of users don't care about the separation, they are just going to see that they have to got to multiple different places to set things. Not to keep bringing up coc, but users coming from coc-land will be used to having both server configuration and client configuration for all LSP related things being done in the same place, in their CocConfig. That's a _huge_ strength, because again a lot of users aren't going to care about the separation of server vs client settings.

Regarding the table in #12333, I'll be honest, I don't understand it as it's presented. The value an type for a config value being checked is nice I guess, but this part I don't fully get:

```
optional_on_set_function` is a function to be called when a value is set for the config.
  --            It is passed the new value
```
So this effect is fired when the value is set or changed? So is this meant to be changed during an active session? Can you provide a real-world example of this table config being used?

**Comment by mfussenegger** [comment]
@landerlo  I think I get your point better now. I saw it too much under the lenses of "strategy for LSP configuration" instead of the more concrete case that you are trying to address there. So my remarks were more on a meta level and probably not useful feedback for the POC PR.


---

I agree also with @ckipp01  that the current split with nvim-lsp is a bit problematic. 

`nvim-jdtls` also needs `init_options` and extended client capabilities for some of the functionality. Currently it looks as if the best option is to delegate that responsibility to the users, which makes using it a bit more difficult.

But I hope this will get better once nvim-lsp focuses more on pure configuration and some of the utilities (like the buffer/client attach manager and some of path functions) move into core.

**Comment by landerlo** [comment]
Yes, I was suspecting I hadn't being clear on the intent and  I completely agree with the meta level approach you outlines. The gist of what I'm proposing for how to have composable building blocks is an extra layering of callbacks: e.g. ```process_location_response_callback(on_locations)``` so the outer callback can do the nitty gritty and sanitize, e.g. for a single location some servers return string and other returns array of one element. So that result is passed to the higher lever callback that only takes a  sanitized array of locations of at least one element tha's why is easy to pass function(locs) lsp.util.jump_to_locations(locs[1]) end. This layering of callbacks is the equivalent of decomposing functions: f(g(c)) but in the async world, and I havent's seen it used and I think it can simplify things a lot. Leverages the integration of the callback system and lua engine, making this viable.

I'm saying stick to protocol and have one 80% callback. If somebody wants to do something different when there are no locations returned then they replace the outer level and do it. I think we should encourage and leverage this calllback composition technique where the outer layer does the nitty gritty of range calculation, etc. and passing a canonical list of elements that can be easily processed in a one liner function that pass them to other utiliy functions.

**Comment by clason** [comment]
@mfussenegger Not to beat a dead horse, but another argument for exposing utility functions (marked as `_private`) is to be able to write tests for them (which otherwise isn't possible with the current infrastructure).

**Comment by tjdevries** [comment]
Inspired by @landerlo 's PR, I have begun work on extending the idea in my PR that was originally just about the config table ( https://github.com/neovim/neovim/pull/12333 ). The end result would allow configuration like the below. The PR is currently still WIP (and names are up for discussion), but I think this snippet gives an example of how easy it could be to chain together multiple, built-in callbacks to provide a very extensible experience for end-users with no copy and paste for small nit picks.

```vim
lua Location = require('vim.lsp.actions').Location

" Jump & then highlight the location
nnoremap <space>gd <cmd>lua vim.lsp.buf.definition { callbacks = { Location.jump_first, Location.highlight } }<CR>

" Preview location
nnoremap <space>pd <cmd>lua vim.lsp.buf.definition { callbacks = { Location.preview } }<CR>
```

This method provides a variety of other benefits as well, but I need to settle on some implementation details before I can describe them easily in comment form.

P.S. I have been writing up a response to several of the discussions above but it is taking a long time and I prefer code as example rather than me trying to explain everything.

**Comment by runiq** [comment]
There have now been multipl requests for the ability to let plugins add support for certain LSP capabilities (#12573: Support for snippets completion items, #12392: Support for `$/progress`).

**Comment by tjdevries** [comment]
This is resolved with `vim.lsp.with` which you can see an example of with `vim.lsp.diagnostic.on_publish_diagnostics()` in the latest master.

**Comment by justinmk** [comment]
`on_func` https://github.com/neovim/neovim/pull/22598 is a proposed generalization of `vim.lsp.with`. Though the ideal solution for LSP extensions is https://github.com/neovim/neovim/pull/22598#issuecomment-1463367606 : 

> * Have some kind of future/promise/async-thunk/whatever (relates a bit to [Lua: structured concurrency, Promises, task pipelines #19624](https://github.com/neovim/neovim/issues/19624)) that each of the `vim.lsp.buf` methods returns that allows to
>   a) cancel
>   b) use the _synthesized_ result of the operation by being able to attach an additional callback or map the result. This would address [feat(lsp): return result of buf_request() in LSP functions #18372](https://github.com/neovim/neovim/pull/18372) and lots of other feature requests.
>   c) await (coroutine aware) or get(blocking) the result.
>   I already have some local version where I played around with this a bit. There I also have a new `vim.lsp.request` that would replace `buf_request`, `buf_request_all` and `buf_request_sync` and also solve the mixed-encoding warnings.
> * Maybe add a `LspResponse` autocmd that's triggered whenever a response is received with `data` set to the payload as a generic listener mechanism.
